kong/api/routes/certificates.lua

```diff
@@ -98,13 +101,129 @@ return {
       })
     end,
 
-
     PUT = function(self, dao_factory, helpers)
-      return crud.put(self.params, dao_factory.ssl_certificates)
+      -- no id present, behaviour should be same as POST
+      if not self.params.id then
+        return create_certificate(self, dao_factory, helpers)
+      end
+
+      --id present in body
+
+      -- check if exists
+      local ssl_cert, err = dao_factory.ssl_certificates:find {
+        id = self.params.id
+      }
+
+      if err then
+        return helpers.yield_error(err)
+      end
+
+      if not ssl_cert then
+        return helpers.responses.send_HTTP_NOT_FOUND()
+      end
+
+      local snis = nil
+      if type(self.params.snis) == "string" then
+        snis = utils.split(self.params.snis, ",")
+      end
+
+      self.params.snis = nil
+
+      local seen = {} -- check for duplicate snis in the request
+      local skip = {} -- avoid db insert if sni is already present in db
+
+      -- if snis field present
+      -- 1. no duplicate snis should be present in the request
+      -- 2. check if any sni in the request is using a cert 
+      --    other than the one being updated
+
+      if snis then
+        for _, sni in ipairs(snis) do
+          if seen[sni] then
+            return helpers.responses.send_HTTP_CONFLICT(
+              "duplicate requested sni name " .. sni
+            )
+          end
+
+          local sni_in_db, err = dao_factory.ssl_servers_names:find({
+            name = sni,
+          })
+          if err then
+            return helpers.yield_error(err)
+          end
+
+          if sni_in_db then
+            if sni_in_db.ssl_certificate_id ~= ssl_cert.id then
+              return helpers.responses.send_HTTP_CONFLICT(
+                "certificate with id " .. sni_in_db.ssl_certificate_id .. 
+                " in use for entry with name " .. sni
+              )
+            end
+            skip[sni] = true
+          end
+          seen[sni] = true
+        end
+      end
+
+      local old_snis, err = dao_factory.ssl_servers_names:find_all {
+        ssl_certificate_id = ssl_cert.id
+      }
+      if err then
+        return helpers.yield_error(err)
+      end
+
+      -- update certificate if necessary
+      if self.params.key and self.params.cert then
```

Comment:
Blocker: If a user sends an incomplete payload (e.g. `id` + `snis` + `key` only), this condition will not evaluate to true, hence not updating the certificate, but the execution will continue to the `snis` logic and potentially return `200 OK` for the update. We could be deceiving the user into thinking their request was successful, when in fact the `key` was not updated.

Since both of those properties are required for an SSL cert, it might be worth not checking for their existence, and instead systematically call `:update(self.params, { id = self.params.id })`. The DAO should return a schema error (`400 Bad Request`) to the client if one of these properties is missing - because we are not specifying the `{ full = true }` option (which prevents the update of `created_at`, iirc).

I see two options:
1. Let the DAO handle such errors and return early with `400 Bad Request` to not run subsequent SNIs logic.
2. Use the `{ full = true }` option and allow for `created_at` updates (minor), but handle such schema validation by hand (that both a `key` and a `cert` are required) and return `400 Bad Request` when one of them is missing.

Of course, tests would be required either way... :D

Reply:
`PATCH` or `PUT` should/do not check for required fields.

Initially, I was trying to use option 1 so I tried the following (**or** instead of **and**)
```lua
if self.params.key or self.params.cert then
        local ssl_cert, err = dao_factory.ssl_certificates:update(self.params,
          {id = self.params.id})
end
```
In such a case, if only `key` or only `cert` is present in the request, DAO does not return a `400 Bad Request`. So, we cannot use option 1 mentioned above.

We can do either of the following:
1. Handle schema validation by hand and perform `:update(self.params, { id = self.params.id })`
2. Use `{ full = true }` and then let DAO handle the error. In this case, I still get `created_at is required`. Do we need to set it before making the update request?

I'm not sure if what you proposed was incorrect or I'm understanding you wrong.

Reply:
Indeed, when passing `{ full = true }`, all required fields must be given. It should be fine since we can get the value of `ssl_cert.created_at` (retrieved earlier in this handler). Here we would let the DAO do its work (and report correctly formatted errors with minimal effort).

The manual validation option is also on the table, but would require some additional duct-tape  which doesn't feel very nice and should be avoided *if possible*.

Could we give a try to `{ full = true }` with the necessary tweaks?